File: game-engine/src/main/java/org/luxons/sevenwonders/game/resources/BestPriceCalculator.java
package org.luxons.sevenwonders.game.resources;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;

public class BestPriceCalculator {

    private final Resources resourcesLeftToPay;

    private final List<ResourcePool> pools;

    private final List<BoughtResources> boughtResources;

    private int pricePaid;

    private List<BoughtResources> bestSolution;

    private int bestPrice;

    private BestPriceCalculator(Resources resourcesToPay, Table table, int playerIndex) {
        Board board = table.getBoard(playerIndex);
        this.resourcesLeftToPay = resourcesToPay.minus(board.getProduction().getFixedResources());
        this.pools = createResourcePools(table, playerIndex);
        this.boughtResources = new ArrayList<>();
        this.pricePaid = 0;
        this.bestSolution = null;
        this.bestPrice = Integer.MAX_VALUE;
    }

    private static List<ResourcePool> createResourcePools(Table table, int playerIndex) {
        Provider[] providers = Provider.values();
        List<ResourcePool> pools = new ArrayList<>(providers.length + 1);

        Board board = table.getBoard(playerIndex);
        TradingRules rules = board.getTradingRules();
        // we only take alternative resources here, because fixed resources were already removed for optimization
        Set<Set<ResourceType>> ownBoardChoices = board.getProduction().getAlternativeResources();
        pools.add(new ResourcePool(null, rules, ownBoardChoices));

        for (Provider provider : providers) {
            Board providerBoard = table.getBoard(playerIndex, provider.getBoardPosition());
            ResourcePool pool = new ResourcePool(provider, rules, providerBoard.getPublicProduction().asChoices());
            pools.add(pool);
        }
        return pools;
    }

    public static int bestPrice(Resources resources, Table table, int playerIndex) {
        BestPriceCalculator bestPriceCalculator = new BestPriceCalculator(resources, table, playerIndex);
        bestPriceCalculator.computePossibilities();
        return bestPriceCalculator.bestPrice;
    }

    public static List<BoughtResources> bestSolution(Resources resources, Table table, int playerIndex) {
        BestPriceCalculator bestPriceCalculator = new BestPriceCalculator(resources, table, playerIndex);
        bestPriceCalculator.computePossibilities();
        return bestPriceCalculator.bestSolution;
    }

    private void computePossibilities() {
        if (resourcesLeftToPay.isEmpty()) {
            updateBestSolutionIfNeeded();
            return;
        }
        for (ResourceType type : ResourceType.values()) {
            if (resourcesLeftToPay.getQuantity(type) > 0) {
                for (ResourcePool pool : pools) {
                    boolean ownResource = pool.getProvider() == null;
                    if (ownResource) {
                        resourcesLeftToPay.remove(type, 1);
                        computePossibilitiesWhenUsing(type, pool);
                        resourcesLeftToPay.add(type, 1);
                        continue;
                    }
                    BoughtResources boughtRes = new BoughtResources(pool.getProvider(), Resources.of(type));
                    int cost = pool.getCost(type);

                    resourcesLeftToPay.remove(type, 1);
                    boughtResources.add(boughtRes);
                    pricePaid += cost;
                    computePossibilitiesWhenUsing(type, pool);
                    pricePaid -= cost;
                    boughtResources.remove(boughtRes);
                    resourcesLeftToPay.add(type, 1);
                }
            }
        }
    }

    private void computePossibilitiesWhenUsing(ResourceType type, ResourcePool pool) {
        for (Set<ResourceType> choice : pool.getChoices()) {
            if (choice.contains(type)) {
                Set<ResourceType> temp = EnumSet.copyOf(choice);
                choice.clear();
                computePossibilities();
                choice.addAll(temp);
            }
        }
    }

    private void updateBestSolutionIfNeeded() {
        if (pricePaid < bestPrice) {
            bestPrice = pricePaid;
            bestSolution = new ArrayList<>(boughtResources);
        }
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/resources/BestPriceCalculator.kt
package org.luxons.sevenwonders.game.resources

import org.luxons.sevenwonders.game.Player
import java.util.ArrayList
import java.util.EnumSet

internal fun bestSolution(resources: Resources, player: Player): TransactionPlan =
    BestPriceCalculator(resources, player).computeBestSolution()

internal data class TransactionPlan(val price: Int, val possibleTransactions: Set<ResourceTransactions>)

private class ResourcePool(
    val provider: Provider?,
    private val rules: TradingRules,
    val choices: Set<MutableSet<ResourceType>>
) {
    fun getCost(type: ResourceType): Int = if (provider == null) 0 else rules.getCost(type, provider)
}

private class BestPriceCalculator(resourcesToPay: Resources, player: Player) {

    private val pools: List<ResourcePool>
    private val resourcesLeftToPay: MutableResources
    private val boughtResources: MutableMap<Provider, MutableResources> = HashMap()
    private var pricePaid: Int = 0

    private var bestSolutions: MutableSet<ResourceTransactions> = mutableSetOf()
    private var bestPrice: Int = Integer.MAX_VALUE

    init {
        val board = player.board
        this.resourcesLeftToPay = resourcesToPay.minus(board.production.getFixedResources()).toMutableResources()
        this.pools = createResourcePools(player)
    }

    private fun createResourcePools(player: Player): List<ResourcePool> {
        val providers = Provider.values()

        val board = player.board
        val rules = board.tradingRules

        val pools = ArrayList<ResourcePool>(providers.size + 1)
        // we only take alternative resources here, because fixed resources were already removed for optimization
        val ownBoardChoices = board.production.getAlternativeResources()
        pools.add(ResourcePool(null, rules, ownBoardChoices.map { it.toMutableSet() }.toSet()))

        for (provider in providers) {
            val providerBoard = player.getBoard(provider.boardPosition)
            val choices = providerBoard.publicProduction.asChoices().map { it.toMutableSet() }.toSet()
            val pool = ResourcePool(provider, rules, choices)
            pools.add(pool)
        }
        return pools
    }

    fun computeBestSolution(): TransactionPlan {
        computePossibilities()
        return TransactionPlan(bestPrice, bestSolutions)
    }

    private fun computePossibilities() {
        if (resourcesLeftToPay.isEmpty()) {
            updateBestSolutionIfNeeded()
            return
        }
        for (type in ResourceType.values()) {
            if (resourcesLeftToPay[type] > 0) {
                for (pool in pools) {
                    if (pool.provider == null) {
                        computeSelfPossibilities(type, pool)
                    } else {
                        computeNeighbourPossibilities(pool, type, pool.provider)
                    }
                }
            }
        }
    }

    private fun computeSelfPossibilities(type: ResourceType, pool: ResourcePool) {
        resourcesLeftToPay.remove(type, 1)
        computePossibilitiesWhenUsing(type, pool)
        resourcesLeftToPay.add(type, 1)
    }

    private fun computeNeighbourPossibilities(pool: ResourcePool, type: ResourceType, provider: Provider) {
        val cost = pool.getCost(type)
        resourcesLeftToPay.remove(type, 1)
        buyOne(provider, type, cost)
        computePossibilitiesWhenUsing(type, pool)
        unbuyOne(provider, type, cost)
        resourcesLeftToPay.add(type, 1)
    }

    fun buyOne(provider: Provider, type: ResourceType, cost: Int) {
        boughtResources.getOrPut(provider) { MutableResources() }.add(type, 1)
        pricePaid += cost
    }

    fun unbuyOne(provider: Provider, type: ResourceType, cost: Int) {
        pricePaid -= cost
        boughtResources.get(provider)!!.remove(type, 1)
    }

    private fun computePossibilitiesWhenUsing(type: ResourceType, pool: ResourcePool) {
        for (choice in pool.choices) {
            if (choice.contains(type)) {
                val temp = EnumSet.copyOf(choice)
                choice.clear()
                computePossibilities()
                choice.addAll(temp)
            }
        }
    }

    private fun updateBestSolutionIfNeeded() {
        if (pricePaid > bestPrice) return

        if (pricePaid < bestPrice) {
            bestPrice = pricePaid
            bestSolutions.clear()
        }
        // avoid mutating the resources from the transactions
        bestSolutions.add(boughtResources.mapValues { MutableResources(HashMap(it.value.quantities)) }.toTransactions())
    }
}
